force-app/main/default/classes/PMT_RefundService.cls

```diff
@@ -218,57 +335,78 @@ public inherited sharing class PMT_RefundService {
-            DebitType__c = (originalPayment.npe01__Payment_Amount__c == refundAmount)
-                ? PMT_RefundService.FULL_REFUND
-                : PMT_RefundService.PARTIAL_REFUND
+    private void processOpportunities() {
+        this.opportunityMap = new Map<Id, Opportunity>(
+            opportunitySelector.getRefundOpportunities(
+                UTIL_SObject.extractId(this.refundRecords, npe01__OppPayment__c.npe01__Opportunity__c))
         );
-        
-        if (isMultiCurrencyEnabled) {
-            refund.put(UTIL_Currency.CURRENCY_ISO_CODE_FIELD, originalPayment.get(UTIL_Currency.CURRENCY_ISO_CODE_FIELD));
-        }
 
-        if (this.refundRecords == null) {
-            this.refundRecords = new List<npe01__OppPayment__c>();
+        for (Integer i = 0; i < this.refundRecords.size(); i++) {
+            Opportunity parentOpp = this.opportunityMap.get(this.refundRecords[i].npe01__Opportunity__c);
+            parentOpp.Amount = (parentOpp.Amount == null) ? 0 : parentOpp.Amount;
+            parentOpp.Amount += this.refundRecords[i].npe01__Payment_Amount__c;
+
+            if (this.oldRefundRecords != null && !this.oldRefundRecords.isEmpty()) {
+                parentOpp.Amount -= this.oldRefundRecords[i].npe01__Payment_Amount__c;
+            }
         }
 
-        this.refundRecords.add(refund);
-        return refund;
     }
 
-    public void updateParentOpportunity(npe01__OppPayment__c originalPayment, Decimal refundAmount) {
-        if (this.opportunityMap == null) {
-            this.opportunityMap = new Map<Id,Opportunity>();
+    private void convertFixedAmountAllocationsToPercentage(List<Allocation__c> allocations) {
+        if (allocations == null || allocations.isEmpty()) {
+            return;
         }
-        Opportunity opportunity = this.opportunityMap.get(originalPayment.npe01__Opportunity__c);
-        if (opportunity == null) {
-            opportunity = new Opportunity(
-                Id = originalPayment.npe01__Opportunity__c,
-                Amount = (originalPayment.npe01__Opportunity__r.Amount == null)
-                    ? 0
-                    : originalPayment.npe01__Opportunity__r.Amount
-            );
+
+        List<Allocation__c> toUpdate = new List<Allocation__c>();
+
+        Decimal oppAmount = allocations[0].Opportunity__r.Amount;
+        Double percentage = 0.0;
+        for (Allocation__c allocation : allocations) {
+            if (allocation.Percent__c == null) {
+                Decimal calculatedPercentage = allocation.Amount__c / oppAmount * 100;
+                allocation.Percent__c = calculatedPercentage.setScale(2, RoundingMode.HALF_UP);
+                toUpdate.add(allocation);
+            }
+
+            percentage += allocation.Percent__c;
         }
 
-        opportunity.Amount = opportunity.Amount - refundAmount;
+        if (percentage > 100) {
+            toUpdate[0].Percent__c = toUpdate[0].Percent__c - (percentage - 100);
```

Comment:
When would this happen, due to rounding? A comment would be helpful, this feels like an odd situation.

Reply:
Correct, this is to handle the edge case of an edge case. If the allocations have some extreme weird setup and we don't want it to be off by 0.01%. I will add the comment in the address it